modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -1858,6 +1862,9 @@ private void processDynamicCacheChangeRequest(DynamicCacheChangeBatch msg) {
             depActions.servicesToUndeploy(toUndeploy);
 
             msg.servicesDeploymentActions(depActions);
+
+            // Remove the names from the service-by-name map as well
```

Comment:
I don't think this is correct. We should only update `registeredServicesByName` right where `registeredServices` is updated, so they are always in sync.

Reply:
Like this?

                registeredServices.entrySet().removeIf(e -> {
                    ServiceInfo desc = e.getValue();

                    if (Objects.equals(desc.cacheName(), chReq.cacheName())) {
                        toUndeploy.put(desc.serviceId(), desc);
                        registeredServicesByName.remove(desc.name());

                        return true;
                    }

                    return false;
                });

Reply:
I mean, wherever we have `registeredServices.put` or `.remove` or `.clear`, we also update `registeredServicesByName`.

On this line there is no such thing, so let's revert it. Other cases are already handled.

Reply:
See the removeIf() line - it's a foreach + remove in one.